embassy-stm32/src/timer/simple_pwm.rs

```diff
@@ -89,26 +194,80 @@ impl<'d, T: GeneralInstance4Channel> SimplePwm<'d, T> {
         this
     }
 
-    /// Enable the given channel.
-    pub fn enable(&mut self, channel: Channel) {
-        self.inner.enable_channel(channel, true);
+    /// Get a single channel
+    ///
+    /// If you need to use multiple channels, use [`Self::split`].
+    pub fn channel(&mut self, channel: Channel) -> SimplePwmChannel<'_, T> {
+        SimplePwmChannel {
+            timer: &self.inner,
+            channel,
+        }
     }
 
-    /// Disable the given channel.
-    pub fn disable(&mut self, channel: Channel) {
-        self.inner.enable_channel(channel, false);
+    /// Channel 1
+    ///
+    /// This is just a convenience wrapper around [`Self::channel`].
+    ///
+    /// If you need to use multiple channels, use [`Self::split`].
+    pub fn ch1(&mut self) -> SimplePwmChannel<'_, T> {
+        self.channel(Channel::Ch1)
     }
 
-    /// Check whether given channel is enabled
-    pub fn is_enabled(&self, channel: Channel) -> bool {
-        self.inner.get_channel_enable_state(channel)
+    /// Channel 2
+    ///
+    /// This is just a convenience wrapper around [`Self::channel`].
+    ///
+    /// If you need to use multiple channels, use [`Self::split`].
+    pub fn ch2(&mut self) -> SimplePwmChannel<'_, T> {
+        self.channel(Channel::Ch2)
+    }
+
+    /// Channel 3
+    ///
+    /// This is just a convenience wrapper around [`Self::channel`].
+    ///
+    /// If you need to use multiple channels, use [`Self::split`].
+    pub fn ch3(&mut self) -> SimplePwmChannel<'_, T> {
+        self.channel(Channel::Ch3)
+    }
+
+    /// Channel 4
+    ///
+    /// This is just a convenience wrapper around [`Self::channel`].
+    ///
+    /// If you need to use multiple channels, use [`Self::split`].
+    pub fn ch4(&mut self) -> SimplePwmChannel<'_, T> {
+        self.channel(Channel::Ch4)
+    }
+
+    /// Splits a [`SimplePwm`] into four pwm channels.
+    ///
+    /// This returns all four channels, including channels that
+    /// aren't configured with a [`PwmPin`].
+    // TODO: I hate the name "split"
+    pub fn split(&mut self) -> SimplePwmChannels<'_, T> {
```

Comment:
with this signature, the split channels still borrow from `self`, so you can't easily create a PWM in main then pass ownership of some channels to TaskA and others to TaskB. You'd have to put the PWM in a StaticCell to get the lifetimes right.

Instead the signature could be `fn split(self) -> SimplePwmChannels<'d, T>`. This way the unsplit PWM is simply gone, and the channels can live independently. It should be doable by unsafely duplicating the `Timer`, or making the channels contain a pointer to PAC regs.

Reply:
I can't seem to unsafely clone the timer PeripheralRef, there is a missing trait bound somewhere.

Aside from that, it is a bit unfortunate that splitting the channels means that the timer will never be disabled, even if all the channels are dropped.

Ideally, the channels would be refcounted with bit flags, which could also make it easier to hand out channels without accidentally duplicating channels. That sounds pretty complex though, so I am going to leave it as a "future possibility" for now.

Reply:
yeah leaving it for future work sounds good to me.